filterExprParser.go

```diff
@@ -0,0 +1,691 @@
+// Condition				= [ "NOT" ] Operand
+// Operand					= { Paren } BooleanExpr | ( LHS ( CheckOp | ( CompareOp RHS) ) ) { Paren }
+// Paren			    	= OpenParen | CloseParen
+// BooleanExpr				= Boolean | BooleanFuncExpr
+// LHS 						= ConstFuncExpr | Field | Value
+// RHS 						= ConstFuncExpr | Value | Field
+// CompareOp				= "=" | "==" | "!=" | "<>" | ">" | ">=" | "<" | "<="
+// CheckOp					= "EXISTS" | ( "IS" [ "NOT" ] ( NULL | MISSING ) )
+// Field					= OnePath { "." OnePath }
+// OnePath					= ( PathFuncExpression | @String | @Ident ){ ArrayIndex }
+// ArrayIndex				= "[" @"-" @Int "]"
+// Value					= "@String"
+// ConstFuncExpr			= ConstFuncNoArg | ConstFuncOneArg | ConstFuncTwoArgs
+// ConstFuncNoArg			= ( ConstFuncNoArgName "(" ")" )
+// ConstFuncNoArgName 		= "PI" | "E"
+// ConstFuncOneArg 			= ( ConstFuncOneArgName "(" ConstFuncArgument ")" )
+// ConstFuncOneArgName  	= "ABS" | "ACOS"...
+// ConstFuncTwoArgs			= ( ConstFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgument ")" )
+// ConstFuncTwoArgsName 	 = "ATAN2" | "POW"
+// ConstFuncArgument		= Field | Value | ConstFuncExpr
+// PathFuncExpression		= OnePathFuncNoArg
+// OnePathFuncNoArg			= ( OnePathFuncNoArgName "(" ")" )
+// OnePathFuncNoArgName		= "META"
+// BooleanFuncExpr			= BooleanFuncTwoArgs
+// BooleanFuncTwoArgs		= ( BooleanFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgument ")" )
+// BooleanFuncTwoArgsName 	= "REGEX_CONTAINS"
+
+type FilterExpression struct {
+	AndConditions []*FEAndCondition `@@ { "OR" @@ }`
+}
+
+func (fe *FilterExpression) String() string {
+	output := []string{}
+
+	first := true
+	for _, expr := range fe.AndConditions {
+		if first {
+			first = false
+		} else {
+			output = append(output, "OR")
+		}
+		output = append(output, expr.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+type FEParen struct {
+	OpenParen  *FEOpenParen  `@@ |`
+	CloseParen *FECloseParen `@@`
+}
+
+func (fep *FEParen) IsOpen() bool {
+	return fep.OpenParen != nil && fep.CloseParen == nil
+}
+
+func (fep *FEParen) IsClose() bool {
+	return fep.CloseParen != nil && fep.OpenParen == nil
+}
+
+func (fep *FEParen) String() string {
+	if fep.IsOpen() {
+		return fep.OpenParen.String()
+	} else if fep.IsClose() {
+		return fep.CloseParen.String()
+	} else {
+		return "?? (FEParen)"
+	}
+}
+
+type FEOpenParen struct {
+	Parens string `@"("`
+}
+
+func (feop *FEOpenParen) String() string {
+	return "("
+}
+
+type FECloseParen struct {
+	Parens string `@")"`
+}
+
+func (fecp *FECloseParen) String() string {
+	return ")"
+}
+
+type FEAndCondition struct {
+	OrConditions []*FECondition `@@ { "AND" @@ }`
+}
+
+func (ac *FEAndCondition) String() string {
+	output := []string{}
+
+	first := true
+	for _, e := range ac.OrConditions {
+		if first {
+			first = false
+		} else {
+			output = append(output, "AND")
+		}
+		output = append(output, e.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+type FECondition struct {
+	PreParen  []*FEParen   `{ @@ }`
```

Comment:
should preParen always be OpenParen and PostParen be CloseParen?

Reply:
Sure. Can do that.